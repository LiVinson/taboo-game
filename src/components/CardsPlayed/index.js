import CardsPlayed from "./CardsPlayed"

export default CardsPlayed