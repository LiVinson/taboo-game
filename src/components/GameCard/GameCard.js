import React from 'react'
import PropType from 'prop-types'
import { ButtonTabooCard, TabooCard } from 'components/shared/TabooCard'
import { InstructionsText, KeyWord } from './style'


export const GiverGameCard = ({ currentCard, changeCardStatus, isPending, error }) => {
	const buttonInfo = [
		{
			text: 'Skip!',
			disabled: isPending,
			onClick: () => {
				changeCardStatus('skipped')
			},
		},
		{
			text: 'Next!',
			disabled: isPending,
			onClick: () => {
				changeCardStatus('correct')
			},
		},
	]


	return <ButtonTabooCard buttons={buttonInfo} tabooWord={currentCard.word} list={currentCard.tabooList} error={error} />
}

GiverGameCard.propType = {
	currentCard: PropType.object.isRequired,
	changeCardStatus: PropType.func.isRequired,
	isPending: PropType.bool.isRequired
}

export const WatcherGameCard = ({currentCard, changeCardStatus, isPending, error}) => {
	const buttonInfo = [
		{
			text: 'Buzzer!',
			disabled: isPending,
			onClick: () => {
				changeCardStatus('discard')
			}
		},
	]

	return <ButtonTabooCard buttons={buttonInfo} tabooWord={currentCard.word} list={currentCard.tabooList} error={error} />
}

WatcherGameCard.propType = {
	currentCard: PropType.object.isRequired,
	changeCardStatus: PropType.func.isRequired,
	isPending: PropType.bool.isRequired
}

export const TeamGameCard = ({ role, giver, watcher }) => {
	return (
		<React.Fragment>
			{role === 'giverTeam' ? (
				<TabooCard tabooWord="Guess!">
					<InstructionsText>
						It's your teams turn to guess the word! <KeyWord>{giver.name}</KeyWord> is giving the clues and{' '}
						<KeyWord>{watcher.name}</KeyWord> will be making sure they don't say anything Taboo!
					</InstructionsText>
				</TabooCard>
			) : (
				<TabooCard tabooWord="Relax!">
					<InstructionsText>
						It’s the other team’s turn to give clues and guess, so don't say anything! <KeyWord>{watcher.name}</KeyWord> will be
						watching to make sure <KeyWord>{giver.name}</KeyWord> doesn’t say any Taboo words.
					</InstructionsText>
				</TabooCard>
			)}
		</React.Fragment>
	)
}

TeamGameCard.propType = {
	role: PropType.string.isRequired,
	giver: PropType.string.isRequired,
	watcher: PropType.string.isRequired,
}

export const GameCard = (props) => {
	const currentCard = props.deck[props.cardIndex]
	switch (props.role) {
		case 'giver':
			return <GiverGameCard {...props} currentCard={currentCard} />
		case 'watcher':
			return <WatcherGameCard {...props} currentCard={currentCard} />
		case 'giverTeam':
		case 'watcherTeam':
			//consider checking if card is changing for animation purposes
			return <TeamGameCard {...props} />
		default:
			return null
	}
}

GameCard.propType = {
	gamecode: PropType.string.isRequired,
	deck: PropType.object.isRequired,
	role: PropType.string.isRequired,
	error: PropType.string,
	cardIndex: PropType.number.isRequired,
	changeCardStatus: PropType.func.isRequired,
	giver: PropType.object.isRequired,
	isPending: PropType.bool.isRequired,
	round: PropType.number.isRequired,
	watcher: PropType.object.isRequired,
}
