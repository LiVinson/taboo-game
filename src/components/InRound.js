import React from "react"
import WordCard from "./WordCard"

export default function InRound({wordInfo, isGiver, isWatcher}){
    if (isGiver || isWatcher) {
        console.log(wordInfo)
        const { word, tabooWords} = wordInfo
      return(
        <div>
            <WordCard
                word={word}
                tabooWords={tabooWords}
            />
            <div>
                <button onClick={()=> console.log("skip it!")}>Skip Word!</button>
                <button onClick={()=> console.log("correct!")}>Correct!</button>
            </div>
        
        </div>
      )
    }  else {
        return <p>Someone else's turn to give/watch</p>
    }  

}