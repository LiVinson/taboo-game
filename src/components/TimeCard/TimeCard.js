import React from 'react'
import PropTypes from 'prop-types'
import moment from 'moment'
import { ProgressBar, Bar, TimeText } from './style'

class TimeCard extends React.Component {
	constructor(props) {
		super(props)
		this.state = {
			timeRemaining: 60,
		}
	}

	//Calculates the difference between current time and time round ends in seconds
	//Updates state with the amount of time remaining every second until 0, then giver
	//calld method to endRound and change to PostRound.
	componentDidMount() {
		const { roundEndTime } = this.props
		let timeRemaining = moment(roundEndTime).diff(moment(), 'second')
		this.intervalId = setInterval(() => {
			if (timeRemaining >= 0) {
				timeRemaining--
				this.setState({
					timeRemaining,
				})
			} else {
				clearInterval(this.intervalId)
				//Only one player needs to update the round for everyone
				if (this.props.role === 'giver') {
					this.props.endRound()
				}
			}
		}, 1000)
	}

	componentWillUnmount() {
		if (this.intervalId) {
			clearInterval(this.intervalId)
		}
	}
	
	render() {
		const width = Math.round((this.state.timeRemaining / 60) * 100)
		return (
			<ProgressBar>
				<Bar width={width}></Bar>
				<TimeText width={width}>
					{this.state.timeRemaining > 0 ? this.state.timeRemaining : "Time's Up!"}
				</TimeText>
			</ProgressBar>
		)
	}
}

TimeCard.propTypes = {
	roundEndTime: PropTypes.number,
	role: PropTypes.string.isRequired,
	endRound: PropTypes.func.isRequired,
}
export default TimeCard
