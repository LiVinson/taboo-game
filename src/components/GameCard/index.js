import { GameCard } from "./GameCard"

export { GameCard }