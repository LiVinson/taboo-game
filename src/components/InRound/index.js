import InRound from "./InRound"

export default InRound