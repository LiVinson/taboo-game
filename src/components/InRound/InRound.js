import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import moment from 'moment'
import TimeCard from 'components/TimeCard'
import LoadingCard from 'components/shared/LoadingCard'
import { ErrorCard } from 'components/shared/ErrorCard'
import { GameCard } from 'components/GameCard'
import { changeCardStatus } from 'store/actions/cardActions'

// export const InRound = ({ roundEndTime, endRound, role, error, ...props }) => {
export class InRound extends React.Component {
	constructor(props) {
		super(props)
		this.state = {
			displayCardsRemaining: false,
		}
	}

	componentDidMount() {
		this.checkCardsRemaining()
	}

	componentDidUpdate(prevProps) {
		//If game is set to end and not previously called, call endRound
		if (prevProps.allCardsPlayed !== this.props.allCardsPlayed && this.props.allCardsPlayed) {
			this.props.endRound()
			//If the card index has changed, check if getting down to last cards
		} else if (prevProps.cardIndex !== this.props.cardIndex) {
			this.checkCardsRemaining()
		}
	}

	checkCardsRemaining = () => {
		const cardsRemaining = this.props.lastCardIndex - this.props.cardIndex
		if ((cardsRemaining === 20 || cardsRemaining === 5) && !this.state.displayCardsRemaining) {
			this.toggleCardsRemaining()
		} else if (cardsRemaining <= -1) {
			this.props.endGame()
		}
	}
	//change display status. If changing to true, call function to change back after a few seconds
	toggleCardsRemaining = () => {
		this.setState(
			{
				displayCardsRemaining: !this.state.displayCardsRemaining,
			},
			() => {
				if (this.state.displayCardsRemaining) {
					this.removeCardsRemaining()
				}
			}
		)
	}

	//Sets display cards false after a few seconds
	removeCardsRemaining = () => {
		this.timeOutId = setTimeout(() => {
			if (this.state.displayCardsRemaining) {
				this.toggleCardsRemaining()
			}
		}, 6000)
	}

	componentWillUnmount() {
		if (this.timeOutId) {
			clearTimeout(this.timeOutId)
		}
	}
	render() {
		const {
			giver,
			watcher,
			role,
			roundEndTime,
			currentCard,
			cardIndex,
			lastCardIndex,
			endRound,
			isPending,
			error,
			changeCardStatus,
		} = this.props
		const cardsRemainingMsg = this.state.displayCardsRemaining
			? `Only ${lastCardIndex - cardIndex} cards remaining!`
			: ''
		return (
			<React.Fragment>
				<TimeCard roundEndTime={roundEndTime} endRound={endRound} role={role} />
				{/*Render GameCard as long as current time is before round ends other wise check for error and render. If no error, display Loading until round status changes*/}
				{moment().isBefore(roundEndTime, 'second') ? (
					<GameCard
						giver={giver}
						watcher={watcher}
						role={role}
						currentCard={currentCard}
						isPending={isPending}
						changeCardStatus={changeCardStatus}
						cardsRemainingMsg={cardsRemainingMsg}
						error={error.cardError}
					/>
				) : error.roundError ? (
					<ErrorCard error={error.roundError} />
				) : (
					<LoadingCard message="Checking cards played" />
				)}
			</React.Fragment>
		)
	}
}

InRound.propTypes = {
	gamecode: PropTypes.string.isRequired,
	giver: PropTypes.object.isRequired,
	watcher: PropTypes.object.isRequired,
	role: PropTypes.string.isRequired,
	round: PropTypes.number.isRequired,
	roundEndTime: PropTypes.number.isRequired,
	currentCard: PropTypes.object.isRequired,
	cardIndex: PropTypes.number.isRequired,
	lastCardIndex: PropTypes.number.isRequired,
	allCardsPlayed: PropTypes.bool.isRequired,
	endRound: PropTypes.func.isRequired,
	isPending: PropTypes.bool.isRequired,
	error: PropTypes.object.isRequired,
	changeCardStatus: PropTypes.func.isRequired,
}

const mapStateToProps = (state) => {
	return {
		isPending: state.cards.pending,
		error: {
			cardError: state.cards.error && state.cards.error.errorMessage,
			roundError: state.round.error && state.round.error.errorMessage,
		},
	}
}

const mapDispatchToProps = (dispatch, prevProps) => {
	//The index of the currently displayed card, current round and gamecode
	const { cardIndex, gamecode, round, giver } = prevProps
	const half = giver.team === 'team 1' ? 'top' : 'bottom'
	return {
		changeCardStatus: (cardStatus) =>
			dispatch(changeCardStatus(gamecode, cardStatus, cardIndex, 'in progress', round, half)),
	}
}

export default connect(mapStateToProps, mapDispatchToProps)(InRound)
