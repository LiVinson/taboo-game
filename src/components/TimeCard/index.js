import TimeCard from "./TimeCard"

export default TimeCard