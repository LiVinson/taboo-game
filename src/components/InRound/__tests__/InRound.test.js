import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { ThemeProvider } from 'styled-components'
import theme from '../../../global-design/theme'
import { InRound } from '../InRound'
import { GameCard } from 'components/GameCard'
import TimeCard from 'components/TimeCard'
import { GiverGameCard, WatcherGameCard, TeamGameCard } from 'components/GameCard/GameCard'
import { deck } from '__fixtures__/deck'

jest.mock('moment', () => {
	return () => jest.requireActual('moment')('2020-09-01T00:00:00.000Z')
})

describe('Inround rendering and functionality', () => {
	const date = new Date()
	const endTime = date.setSeconds(date.getSeconds() + 30)
	const defaultProps = {
		gamecode: '90210',
		giver: { name: 'Sam', playerId: '12345', team: 'team 1' },
		watcher: { name: 'Jo', playerId: '23456', team: 'team 2' },
		role: 'giver',
		round: 1,
		roundEndTime: endTime,
		currentCard: deck[0],
		cardIndex: 0,
		lastCardIndex: 10,
		allCardsPlayed: false,
		endRound: jest.fn(),
		isPending: false,
		error: {
			cardError: null,
			roundError: null,
		},
		changeCardStatus: jest.fn(),
	}

	test('renders correctly', () => {
		const props = {
			...defaultProps,
		}

		const wrapper = renderer
			.create(
				<ThemeProvider theme={theme}>
					<InRound {...props} />
				</ThemeProvider>
			)
			.toJSON()
		expect(wrapper).toMatchSnapshot()
	})

	test('renders TimeCard and GiverGame card when role is giver', () => {
		const props = {
			...defaultProps,
		}

		const wrapper = shallow(<InRound {...props} />)

		expect(wrapper.find(TimeCard)).toHaveLength(1)
		expect(wrapper.find(GameCard)).toHaveLength(1)
		expect(wrapper.find(GameCard).dive().find(GiverGameCard)).toHaveLength(1)
		expect(wrapper.find(GameCard).dive().find(WatcherGameCard)).toHaveLength(0)
		expect(wrapper.find(GameCard).dive().find(TeamGameCard)).toHaveLength(0)
	})

	test('renders TimeCard and WatcherGameCard when role is watcher', () => {
		const props = {
			...defaultProps,
			role: 'watcher',
		}

		const wrapper = shallow(<InRound {...props} />)

		expect(wrapper.find(TimeCard)).toHaveLength(1)
		expect(wrapper.find(GameCard)).toHaveLength(1)
		expect(wrapper.find(GameCard).dive().find(GiverGameCard)).toHaveLength(0)
		expect(wrapper.find(GameCard).dive().find(WatcherGameCard)).toHaveLength(1)
		expect(wrapper.find(GameCard).dive().find(TeamGameCard)).toHaveLength(0)
	})

	test('renders TimeCard and TeamGameCard when role is giverTeam', () => {
		const props = {
			...defaultProps,
			role: 'giverTeam',
		}

		const wrapper = shallow(<InRound {...props} />)

		expect(wrapper.find(TimeCard)).toHaveLength(1)
		expect(wrapper.find(GameCard)).toHaveLength(1)
		expect(wrapper.find(GameCard).shallow().find(GiverGameCard)).toHaveLength(0)
		expect(wrapper.find(GameCard).shallow().find(WatcherGameCard)).toHaveLength(0)
		expect(wrapper.find(GameCard).shallow().find(TeamGameCard)).toHaveLength(1)
	})

	test('renders TimeCard and TeamGameCard when role is watcherTeam', () => {
		const props = {
			...defaultProps,
			role: 'watcherTeam',
		}

		const wrapper = shallow(<InRound {...props} />)

		expect(wrapper.find(TimeCard)).toHaveLength(1)
		expect(wrapper.find(GameCard)).toHaveLength(1)
		expect(wrapper.find(GameCard).dive().find(GiverGameCard)).toHaveLength(0)
		expect(wrapper.find(GameCard).dive().find(WatcherGameCard)).toHaveLength(0)
		expect(wrapper.find(GameCard).dive().find(TeamGameCard)).toHaveLength(1)
	})
})
