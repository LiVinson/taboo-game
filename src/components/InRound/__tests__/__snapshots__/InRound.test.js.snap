// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inround rendering and functionality renders correctly 1`] = `
Array [
  <div
    className="sc-AxjAm gtWhgV"
  >
    <div
      className="sc-AxirZ lkJSxu"
      width={100}
    />
    <div
      className="sc-AxiKw hKElRV"
      width={100}
    >
      60
    </div>
  </div>,
  <div
    className="sc-AxhCb gXuWDR"
  >
    <div
      className="sc-AxgMl fZxUmv"
    >
      apple
    </div>
    <div
      className="sc-AxheI gIsRbv"
    >
      <ul
        className="sc-fzoLsD jOQREE"
      >
        <li
          className="sc-Axmtr eKdZYh"
        >
          fruit
        </li>
        <li
          className="sc-Axmtr eKdZYh"
        >
          red
        </li>
        <li
          className="sc-Axmtr eKdZYh"
        >
          shiny
        </li>
        <li
          className="sc-Axmtr eKdZYh"
        >
          round
        </li>
        <li
          className="sc-Axmtr eKdZYh"
        >
          Eve
        </li>
      </ul>
    </div>
    <div
      className="sc-fznZeY heqlvS"
    >
      <button
        className="sc-fzpans cGkFOO"
        disabled={false}
        onClick={[Function]}
      >
        Skip!
      </button>
      <button
        className="sc-fzpans sc-fzplWN fzsIdO"
        disabled={false}
        onClick={[Function]}
      >
        Next!
      </button>
    </div>
  </div>,
]
`;
